### PR TITLE
More improvements to the Ja Rule admin tool

### PR DESCRIPTION
**Pull request by nomis52**
This adds various types of DMX frames (0-length, back-to-back) as well as basic RDM DUB support.

### CODE REVIEW

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 1]
```diff
@@ -103,19 +102,11 @@ OpenLightingDevice::~OpenLightingDevice() {
...
-  while (true) {
-    MutexLocker locker(&m_in_mutex);
-    if (!m_in_in_progress) {
-      break;
-    }
+  bool transfers_pending = true;
+  while (transfers_pending) {
+    // Spin waiting for the transfers to complete.
+    MutexLocker locker(&m_mutex);
+    transfers_pending = m_out_in_progress || m_in_in_progress;
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: Round brackets please.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 2]
```diff
@@ -140,10 +131,11 @@ bool OpenLightingDevice::Init() {
     return false;
   }
 
-  r = libusb_claim_interface(m_handle, 0);
+  r = libusb_claim_interface(m_handle, 2);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: Can we have this as a const.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 3]
```diff
@@ -157,21 +149,88 @@ bool OpenLightingDevice::SendMessage(Command command,
...
+  MutexLocker locker(&m_mutex);
+  m_in_in_progress = false;
+  m_pending_requests--;
+  if (m_pending_requests) {
+    SubmitInTransfer();
+  }
+}
+
+void OpenLightingDevice::MaybeSendRequest() {
+  if (m_out_in_progress || m_pending_requests > MAX_IN_FLIGHT ||
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: Round brackets around the > please.

File: tools/ja-rule/OpenLightingDevice.h [hunk 4]
```diff
@@ -135,24 +139,34 @@ class OpenLightingDevice {
...
   libusb_device_handle* m_handle;
   MessageHandlerInterface* m_message_handler;
 
-  ola::thread::Mutex m_out_mutex;
-  uint8_t m_out_buffer[OUT_BUFFER_SIZE];  // GUARDED_BY(m_out_mutex);
-  libusb_transfer *m_out_transfer;  // GUARDED_BY(m_out_mutex);
-  bool m_out_in_progress;  // GUARDED_BY(m_out_mutex);
+  ola::thread::Mutex m_mutex;
+  std::queue<PendingRequest> m_queued_requests;  // GUARDED_BY(m_mutex);
+  // The number of request frames we've already sent to the device. We limit
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.h]: SPaG "We limit the number of oustanding requests to MAX_IN_FLIGHT" I think.

File: tools/ja-rule/ja-rule.cpp [hunk 5]
```diff
@@ -159,7 +190,9 @@ class InputHandler {
         m_stdin_handler(
             new StdinHandler(ss, ola::NewCallback(this, &InputHandler::Input))),
         m_device(NULL),
-        m_log_count(0) {
+        m_log_count(0),
+        m_dmx_slot_data(0),
+        m_our_uid(ola::OPEN_LIGHTING_ESTA_CODE, 10) {
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Fixed UID? Add a todo at least.

File: tools/ja-rule/ja-rule.cpp [hunk 6]
```diff
@@ -248,14 +311,38 @@ class InputHandler {
...
+
+  void SendDoubleDMX() {
+    uint8_t payload[512];
+    memset(payload, 0, 512);
+    payload[0] = m_dmx_slot_data;
+    _SendDMX(payload, arraysize(payload));
+    m_dmx_slot_data += 16;
+    payload[0] = m_dmx_slot_data;
+    _SendDMX(payload, arraysize(payload));
+    m_dmx_slot_data += 16;
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Will this wrap cleanly if you keep sending it?

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 1 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Not for this, it's a simple conjunction

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 2 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done.

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 3 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Again it's simple enough it doesn't need it.

File: tools/ja-rule/OpenLightingDevice.h (hunk 4 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.h]: Fixed.

File: tools/ja-rule/ja-rule.cpp (hunk 5 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 6 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: Yes, it's a uint8_t

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 1 above)
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: Okay.

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 3 above)
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: Okay.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 7]
```diff
@@ -140,10 +131,11 @@ bool OpenLightingDevice::Init() {
     return false;
   }
 
-  r = libusb_claim_interface(m_handle, 0);
+  r = libusb_claim_interface(m_handle, INTERFACE_OFFSET);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: It's not really an offset is it, more the interface value for control?

**Comment by peternewman** [comment]
Just one comment.

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 7 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Well on the device side it's an offset. You specify the number interfaces each function driver uses and it computes the value for you.

Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: But not on the machine side.

**Pull request merged**